File: src/graph_mcp/tools/message_tools.py
from __future__ import annotations

import re
from typing import Any

from markdown import markdown

_HTML_TAG_RE = re.compile(r"<\s*[a-zA-Z][^>]*>")
_MARKDOWN_EXTENSIONS = ["extra", "sane_lists", "nl2br"]


def _looks_like_html(text: str) -> bool:
    return bool(_HTML_TAG_RE.search(text))


def markdown_to_html(message: str) -> str:
    """Convert markdown into Teams-friendly HTML via Python-Markdown."""
    if not message.strip():
        return ""

    return markdown(message, extensions=_MARKDOWN_EXTENSIONS)


def normalize_mentions(mentions: list[dict[str, Any]] | None) -> list[dict[str, Any]]:
    if not mentions:
        return []

    normalized: list[dict[str, Any]] = []
    for index, mention in enumerate(mentions):
        if "mentioned" in mention and "mentionText" in mention:
            entry = dict(mention)
            entry.setdefault("id", index)
            normalized.append(entry)
            continue

        raw_id = mention.get("id")
        mention_id = mention.get("mention_id")
        if mention_id is None:
            mention_id = raw_id if isinstance(raw_id, int) else index

        mention_text = (
            mention.get("name")
            or mention.get("display_name")
            or mention.get("displayName")
            or mention.get("mentionText")
            or mention.get("text")
        )
        user_id = (
            mention.get("user_id")
            or mention.get("userId")
            or (raw_id if isinstance(raw_id, str) else None)
        )
        user_identity_type = (
            mention.get("user_identity_type")
            or mention.get("userIdentityType")
            or "aadUser"
        )

        if mention_text is None or user_id is None:
            raise ValueError(
                "Each mention must include either raw Graph mention fields or "
                "a simplified shape with `name`/`display_name` and `user_id`."
            )

        normalized.append(
            {
                "id": mention_id,
                "mentionText": mention_text,
                "mentioned": {
                    "user": {
                        "id": user_id,
                        "displayName": mention_text,
                        "userIdentityType": user_identity_type,
                    }
                },
            }
        )

    return normalized


def build_rich_text_body(
    message: str,
    is_html: bool = True,
    *,
    html_content_type: str = "html",
    text_content_type: str = "text",
) -> dict[str, Any]:
    if is_html:
        content = message if _looks_like_html(message) else markdown_to_html(message)
        content_type = html_content_type
    else:
        content = message
        content_type = text_content_type

    return {"contentType": content_type, "content": content}


def build_chat_message_payload(
    message: str,
    is_html: bool = True,
    mentions: list[dict[str, Any]] | None = None,
) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "body": build_rich_text_body(message, is_html)
    }

    normalized_mentions = normalize_mentions(mentions)
    if normalized_mentions:
        payload["mentions"] = normalized_mentions

    return payload

File: src/tools/message-tools.ts
const INVALID_MENTION_MESSAGE =
  "Each mention must include either raw Graph mention fields or a simplified shape with `name`/`display_name` and `user_id`.";

export type MentionInput = Readonly<Record<string, unknown>>;
export type GraphMention = Record<string, unknown>;

export interface RichTextBody {
  readonly contentType: string;
  readonly content: string;
}

export interface RichTextBodyOptions {
  readonly htmlContentType?: string;
  readonly textContentType?: string;
}

export type ChatMessageImportance = "normal" | "high" | "urgent";

export interface ChatMessageOptions {
  readonly importance?: ChatMessageImportance;
  readonly subject?: string;
}

export interface ChatMessagePayload {
  readonly body: RichTextBody;
  readonly mentions?: GraphMention[];
  readonly importance?: ChatMessageImportance;
  readonly subject?: string;
}

export function normalizeMentions(
  mentions: readonly MentionInput[] | null | undefined,
): GraphMention[] {
  if (mentions === null || mentions === undefined || mentions.length === 0) {
    return [];
  }

  return mentions.map((mention, index) => {
    if ("mentioned" in mention && "mentionText" in mention) {
      const normalized = { ...mention };
      if (!("id" in normalized)) {
        normalized.id = index;
      }
      return normalized;
    }

    const rawId = mention.id;
    const mentionId =
      mention.mention_id !== null && mention.mention_id !== undefined
        ? mention.mention_id
        : typeof rawId === "number" && Number.isInteger(rawId)
          ? rawId
          : index;
    const mentionText =
      mention.name ||
      mention.display_name ||
      mention.displayName ||
      mention.mentionText ||
      mention.text;
    const userId =
      mention.user_id || mention.userId || (typeof rawId === "string" ? rawId : undefined);
    const userIdentityType = mention.user_identity_type || mention.userIdentityType || "aadUser";

    if (
      mentionText === null ||
      mentionText === undefined ||
      userId === null ||
      userId === undefined
    ) {
      throw new Error(INVALID_MENTION_MESSAGE);
    }

    return {
      id: mentionId,
      mentionText,
      mentioned: {
        user: {
          id: userId,
          displayName: mentionText,
          userIdentityType,
        },
      },
    };
  });
}

export function buildRichTextBody(
  message: string,
  isHtml = true,
  options: RichTextBodyOptions = {},
): RichTextBody {
  return {
    contentType: isHtml ? (options.htmlContentType ?? "html") : (options.textContentType ?? "text"),
    content: message,
  };
}

export function buildChatMessagePayload(
  message: string,
  isHtml = true,
  mentions: readonly MentionInput[] | null | undefined = undefined,
  options: ChatMessageOptions = {},
): ChatMessagePayload {
  const normalizedMentions = normalizeMentions(mentions);
  const importance = options.importance ?? "normal";
  const subject = options.subject ?? "";

  return {
    body: buildRichTextBody(message, isHtml),
    ...(normalizedMentions.length > 0 ? { mentions: normalizedMentions } : {}),
    ...(importance === "normal" ? {} : { importance }),
    ...(subject === "" ? {} : { subject }),
  };
}
